### PR TITLE
resolve loader_path

### PR DESCRIPTION
**Pull request by isuruf**

### CODE REVIEW

**Comment by matthew-brett** [comment]
How did `@loader_path` get into the Numpy wheels?

We need some tests...

Does the Numpy wheel work once this patch is applied?

**Comment by isuruf** [comment]
> How did @loader_path get into the Numpy wheels?

It didn't get into the numpy wheels. It's in `libgfortran.dylib` from the `gfortran-darwin-arm64` toolchain.

> Does the Numpy wheel work once this patch is applied?

It should. I haven't checked it.

> We need some tests...

I've no idea how to write a test. Any suggestions?

**Comment by matthew-brett** [comment]
Aha - I don't we had previous seen `@loader_path` inside an `rpath`. 

For the tests - I'm afraid it's a little bit fiddly.  But you could do a first pass by extending the tests in `test_add_rpath` from `delocate/tests/test_install_names.py`...

**Comment by matthew-brett** [comment]
Nice - thanks.  Just to do the final sign-off - would you mind checking if the Numpy wheels work with this fix?

**Comment by isuruf** [comment]
It seems like we are running this function after copying `libgfortran.dylib` to the wheel and therefore resolving `@loader_path` doesn't give the intended path.

File: delocate/libsana.py [hunk 1]
```diff
@@ -13,7 +13,7 @@
 from .tmpdirs import TemporaryDirectory
 
 
-def tree_libs(start_path, filt_func=None):
+def tree_libs(start_path, filt_func=None, original_paths=[]):
```
Comment (matthew-brett) [inline delocate/libsana.py:16]: Avoid mutable default arguments.  Can this be a tuple?

File: delocate/tests/test_install_names.py [hunk 2]
```diff
@@ -141,6 +141,11 @@ def test_add_rpath():
         assert_equal(get_rpaths(libfoo), ('/a/path',))
         add_rpath(libfoo, '/another/path')
         assert_equal(get_rpaths(libfoo), ('/a/path', '/another/path'))
+        add_rpath(libfoo, '@loader_path')
+        assert_equal(get_rpaths(libfoo), ('/a/path', '/another/path', tmpdir))
+        add_rpath(libfoo, '@loader_path/lib')
```
Comment (matthew-brett) [inline delocate/tests/test_install_names.py:146]: Is this now testing the original path manipulation?

File: delocate/libsana.py [hunk 3]
```diff
@@ -56,7 +58,16 @@ def tree_libs(start_path, filt_func=None):
             depending_libpath = realpath(pjoin(dirpath, base))
             if filt_func is not None and not filt_func(depending_libpath):
                 continue
-            rpaths = get_rpaths(depending_libpath)
+
+            depending_libpath_orig = depending_libpath
```
Comment (matthew-brett) [inline delocate/libsana.py:62]: Comments would help here.

File: delocate/delocating.py [hunk 4]
```diff
@@ -199,7 +199,7 @@ def _copy_required(lib_path, copy_filt_func, copied_libs):
     ``dependings_dict`` entry for ``copied_libs['/sys/libB.dylib']``
     """
     # Paths will be prepended with `lib_path`
-    lib_dict = tree_libs(lib_path)
+    lib_dict = tree_libs(lib_path, original_paths=list(copied_libs.keys()))
```
Comment (matthew-brett) [inline delocate/delocating.py:202]: Surely fine as `tuple(copied_libs)`?

**Review by matthew-brett — COMMENTED** [review]
First pass thoughts. I need to reflect a bit on the new argument.

**Comment by isuruf** [comment]
Ping on this

File: delocate/libsana.py [hunk 5]
```diff
@@ -56,7 +58,21 @@ def tree_libs(start_path, filt_func=None):
             depending_libpath = realpath(pjoin(dirpath, base))
             if filt_func is not None and not filt_func(depending_libpath):
                 continue
-            rpaths = get_rpaths(depending_libpath)
+
+            # Sometime rpaths might have @loader_path which means dependent
```
Comment (radarhere) [inline delocate/libsana.py:62]:
```suggestion
            # Sometimes rpaths might have @loader_path which means dependent
```

File: delocate/tools.py [hunk 6]
```diff
@@ -180,6 +180,9 @@ def _cmd_out_err(cmd):
     return out.split('\n')
 
 
+_LINE0_RE = re.compile(r"^(?: \(architecture .*\))?:(?P<further_report>.*)")
```
Comment (matthew-brett) [inline delocate/tools.py:183]: Would you mind splitting these changes off into a separate PR, after #94 is merged?   Maybe add a comment above this line explaining the "architecture" string?

Comment (isuruf) [reply delocate/tools.py:183]: Sure. I'll send a PR after #94 is merged.

**Comment by matthew-brett** [comment]
Can we close this one now #94 is merged and you've factored out #115 ?

**Comment by isuruf** [comment]
Yes

**Pull request closed without merging**